### PR TITLE
Java Implementation for Haybale Stacking + New Notes

### PR DESCRIPTION
**Pull request by brad182**
Added Java implementation that follows the C++ version that is already there, added notes on both the C++ version and the new Java version to help people understand better. I made sure that the implementation for the new Java version is the same as the C++ code, it compiles and passes on Sphere.

### CODE REVIEW

**Review by HarryWangATX — CHANGES_REQUESTED** [review]
Use Kattio pls.

File: solutions/usaco-104.mdx [hunk 1]
```diff
@@ -21,26 +21,67 @@ int main () {
...
+	sort (sol, sol + n); // sort to get median
+	cout << sol[n/2] << '\n'; // output the median
 }
 
 ```
 
 </CPPSection>
+	
 </LanguageSection>
+<JavaSection>
```
Comment (maggie-j-liu) [inline solutions/usaco-104.mdx:48]: this should go inside the LanugageSection; something like:
```
<LanguageSection>

<CPPSection>

c++ code...

</CPPSection>

<JavaSection>

java code...

</JavaSection>

</LanguageSection>
```

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

**Comment by brad182** [comment]
> Use Kattio pls.

used that, is there anything else that needs to be changed?

File: solutions/usaco-104.mdx [hunk 2]
```diff
@@ -21,26 +21,72 @@ int main () {
...
+        }
+
+        int[] prefixSum = new int[n + 1];
+        int total = 0;
+        for (int x = 0; x < n; x++) { // build prefixSum array
+            total += differenceArray[x];
+            prefixSum[x] = total;
+        }
+        Arrays.sort(prefixSum);
+        System.out.println(prefixSum[prefixSum.length / 2]);
```
Comment (HarryWangATX) [inline solutions/usaco-104.mdx:80]:
```suggestion
        io.println(prefixSum[prefixSum.length / 2]);
```

Comment (HarryWangATX) [reply solutions/usaco-104.mdx:80]: and add io.close() after this

File: solutions/usaco-104.mdx [hunk 3]
```diff
@@ -21,26 +21,71 @@ int main () {
...
+
+        int n = io.nextInt();
+        int k = io.nextInt();
+        int[] differenceArray = new int[n + 1]; // differenceArray[n] is the difference between A[n + 1] and A[n]
+
+        for (int x = 0; x < k; x++) { // read input and build differenceArray
+            int a = io.nextInt();
+            int b = io.nextInt();
+            a--; // make zero-indexed
+            b--;
```
Comment (HarryWangATX) [inline solutions/usaco-104.mdx:68]: why not combine into instantiation

**Review by HarryWangATX — CHANGES_REQUESTED** [review]

File: solutions/usaco-104.mdx [hunk 4]
```diff
@@ -21,26 +21,71 @@ int main () {
...
+	sort (sol, sol + n); // sort to get median
+	cout << sol[n/2] << '\n'; // output the median
 }
 
 ```
 
 </CPPSection>
+	
 </LanguageSection>
+<JavaSection>
```
Comment (HarryWangATX) [inline solutions/usaco-104.mdx:48]: why is this under language section?

**Review by HarryWangATX — CHANGES_REQUESTED** [review]

File: solutions/usaco-104.mdx (hunk 1 above)
Comment (maggie-j-liu) [reply solutions/usaco-104.mdx:48]: this is not resolved; move the javasection inside languagesection

File: solutions/usaco-104.mdx [hunk 5]
```diff
@@ -21,26 +21,67 @@ int main () {
...
+
+## Java Implementation
+
+```java
```
[the remark this answers is not in the document]
Comment (maggie-j-liu) [reply solutions/usaco-104.mdx:52]: also not resolved (use tabs instead of spaces to indent)

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/usaco-104.mdx (hunk 1 above)
Comment (brad182) [reply solutions/usaco-104.mdx:47]: is it correct now?

File: solutions/usaco-104.mdx (hunk 3 above)
Comment (brad182) [reply solutions/usaco-104.mdx:67]: i'm not sure what that is, and google searches don't tell me much either

Comment (maggie-j-liu) [reply solutions/usaco-104.mdx:67]: just do 
```
int a = io.nextInt() - 1;
int b = io.nextInt() - 1;
```

File: solutions/usaco-104.mdx [hunk 6]
```diff
@@ -21,26 +21,70 @@ int main () {
...
+    	}
+
+	// CodeSnip{Kattio}
+}
+
+```
+
+</JavaSection>
+	
+<LanguageSection>
```
Comment (maggie-j-liu) [inline solutions/usaco-104.mdx:88]:
```suggestion
</LanguageSection>
```

File: solutions/usaco-104.mdx (hunk 3 above)
Comment (brad182) [reply solutions/usaco-104.mdx]: ok i did that and did i do the other changes right?

sry first time doing this

File: solutions/usaco-104.mdx (hunk 5 above)
[the remark this answers is not in the document]
Comment (maggie-j-liu) [reply solutions/usaco-104.mdx:51]: i think you tried to use tabs, but it didn't work...
an easy way to convert from spaces to tabs is to use [editor](https://usaco.guide/editor/) > paste your code > right click > command palette > convert indentation to tabs

**Comment by brad182** [comment]
i changed the indents to tabs and compared the overall formatting with other approved code, hopefully its right now?

File: solutions/usaco-104.mdx [hunk 7]
```diff
@@ -21,26 +21,71 @@ int main () {
...
+import java.util.*;
+import java.io.*;
+
+class haybaleStacking {
+	public static void main (String[] args) {
+        	Kattio io = new Kattio();
+
+        	int n = io.nextInt();
+        	int k = io.nextInt();
+        	int[] differenceArray = new int[n + 1]; // differenceArray[n] is the difference between A[n + 1] and A[n]
```
Comment (maggie-j-liu) [inline solutions/usaco-104.mdx:61]: this line is too long, could you put the comment above?
```suggestion
		// differenceArray[n] is the difference between A[n + 1] and A[n]
        	int[] differenceArray = new int[n + 1]; 
```

**Comment by maggie-j-liu** [comment]
i think you missed https://github.com/cpinitiative/usaco-guide/pull/1635#discussion_r642184469 (fixing this should fix the build)

some of the indents are tabs but others are still spaces for me? 
image
could you try to convert them again, or i could fix it for you?

everything else looks good (except the change requested above^), but i'll fix some things about the c++ code

**Comment by brad182** [comment]
> i think you missed [#1635 (comment)](https://github.com/cpinitiative/usaco-guide/pull/1635#discussion_r642184469) (fixing this should fix the build)
> 
> some of the indents are tabs but others are still spaces for me?
> image
> could you try to convert them again, or i could fix it for you?
> 
> everything else looks good (except the change requested above^), but i'll fix some things about the c++ code

now on the editor it shows up as all tabs
Screen Shot 2021-05-30 at 8 36 58 PM

**Comment by brad182** [comment]
all issues fixed and ready for approval? or is there something else?

**Comment by maggie-j-liu** [comment]
Please fix https://github.com/cpinitiative/usaco-guide/pull/1635#discussion_r642184469 and https://github.com/cpinitiative/usaco-guide/pull/1635#discussion_r642188447

File: solutions/usaco-104.mdx [hunk 8]
```diff
@@ -21,26 +21,74 @@ int main () {
...
 </CPPSection>
+	
+<JavaSection>
+
+## Java Implementation
+
+```java
+import java.util.*;
+import java.io.*;
+
```
Comment (maggie-j-liu) [inline solutions/usaco-104.mdx:54]:
```suggestion
```

File: solutions/usaco-104.mdx [hunk 9]
```diff
@@ -21,26 +21,74 @@ int main () {
...
+		int total = 0;
+			      
+		for (int x = 0; x < n; x++) { // build prefixSum array
+			total += differenceArray[x];
+			prefixSum[x] = total;
+			}
+
+		Arrays.sort(prefixSum);
+		io.println(prefixSum[prefixSum.length / 2]);
+		}
```
Comment (maggie-j-liu) [inline solutions/usaco-104.mdx:84]:
```suggestion

		for (int x = 0; x < n; x++) { // build prefixSum array
			total += differenceArray[x];
			prefixSum[x] = total;
		}

		Arrays.sort(prefixSum);
		io.println(prefixSum[prefixSum.length / 2]);
	}
```

File: solutions/usaco-104.mdx [hunk 10]
```diff
@@ -21,26 +21,74 @@ int main () {
...
+		int[] prefixSum = new int[n + 1];
+		int total = 0;
+			      
+		for (int x = 0; x < n; x++) { // build prefixSum array
+			total += differenceArray[x];
+			prefixSum[x] = total;
+			}
+
+		Arrays.sort(prefixSum);
+		io.println(prefixSum[prefixSum.length / 2]);
```
Comment (HarryWangATX) [inline solutions/usaco-104.mdx:83]: Where is the `io.close();` I told you to add?

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

**Review by HarryWangATX — CHANGES_REQUESTED** [review]

File: solutions/usaco-104.mdx (hunk 10 above)
Comment (brad182) [reply solutions/usaco-104.mdx:80]: it is in the latest version of the code

**Comment by brad182** [comment]
ok i made all lines less than 80

**Comment by maggie-j-liu** [comment]
i think it's good after you address https://github.com/cpinitiative/usaco-guide/pull/1635#discussion_r642197333

**Review by brad182 — COMMENTED** [review]
> i think it's good after you address [#1635 (comment)](https://github.com/cpinitiative/usaco-guide/pull/1635#discussion_r642197333)

oh lol i did that but it didn't save that time got it now
again sry this is my first time contributing

File: solutions/usaco-104.mdx [hunk 11]
```diff
@@ -2,15 +2,15 @@
...
 source: USACO Bronze 2012 January
 title: Haybale Stacking
-author: Óscar Garries
+author: Óscar Garries, Brad Ma
 ---
 
 [Official Analysis](http://www.usaco.org/current/data/sol_stacking.html)
 
-<LanguageSection>
+</LanguageSection>
```
Comment (maggie-j-liu) [inline solutions/usaco-104.mdx:10]:
```suggestion
<LanguageSection>
```
opening tags don't have `/`, closing tags do

Comment (brad182) [reply solutions/usaco-104.mdx]: rip saw a suggestion for that so I just added the slash oops

File: solutions/usaco-104.mdx [hunk 12]
```diff
@@ -21,26 +21,71 @@ int main () {
...
+		int k = io.nextInt();
+		int[] differenceArray = new int[n + 1]; 
+		// differenceArray[n] is the difference between A[n + 1] and A[n]
+
+		for (int x = 0; x < k; x++) { // read input and build differenceArray
+			int a = io.nextInt() - 1; // read and make zero-indexed
+			int b = io.nextInt() - 1;
+
+			differenceArray[a] = differenceArray[a] + 1;
+			differenceArray[b + 1] = differenceArray[b + 1] - 1;
```
Comment (HarryWangATX) [inline solutions/usaco-104.mdx:68]:
```suggestion
			differenceArray[a]++;
			differenceArray[b + 1]--;
```

**Review by HarryWangATX — CHANGES_REQUESTED** [review]

File: solutions/usaco-104.mdx (hunk 12 above)
Comment (brad182) [reply solutions/usaco-104.mdx]: changed

**Review by maggie-j-liu — APPROVED** [review]

**Review by HarryWangATX — APPROVED** [review]

**Pull request merged**